### PR TITLE
Improve performance of generating dot-bracket strings with pseudoknots

### PR DESCRIPTION
**Pull request by luxaritas**
## Summary
This replaces the implementation of SecStruct.getParenthesis with a more efficient implementation. This also improves the efficiency of filtering for only/no pseudoknots, which is one of the most expensive parts of layout recomputation.

## Implementation Notes
Instead of turning a pair map into stems, then for each stem iterating over all bases between the start and end of the pair to find if a half-pair has previously been placed there, we now walk over each base only once, keeping track of any half-open pairs that have not been closed off yet. This should bring our time complexity from O(n^2) (both in the stem construction and in the string construction) to O(n). Doing a quick and dirty benchmark, I saw up to a 10x performance improvement in this method. It was slower in some cases, but those situation should already be very fast.

In earlier commits of the PR, I first simplified the existing implementation to wrap my head around it a bit better.

## Testing
Added new unit tests, verifying against the previous implementation. As part of my benchmarking experimentation I also validated against the Ribonanza PDB test set.

## Related Issues
Surfaced during work on #748

### CODE REVIEW

**Comment by luxaritas** [comment]
@tkaragianes Beyond just high-level checking my work, hoping you can check if my comments are sufficient to make it clear how and why this works, and whether the test cases are sufficient

**Review by tkaragianes — APPROVED** [review]
Reviewed synchronously with Jonathan. Discussed maybe needing different wording on the comments?

File: src/eterna/rnatypes/SecStruct.ts [hunk 1]
```diff
@@ -333,82 +333,63 @@ export default class SecStruct {
...
-                        const j = stem[jj][1];
 
-                        dbn[i] = charsL[pkCtr];
-                        dbn[j] = charsR[pkCtr];
+            // For each pseudoknot degree/character, we maintain an array of base indices
+            // which are the closing end of a pair (specifically, pairs which it's possible
+            // for us to "cross" with a new pair)
+            const closingBasesPerDegree: number[][] = new Array(charsL.length).fill(null).map(() => []);
+
+            // Walk through all openinng-halves of pairs
```
Comment (tkaragianes) [inline src/eterna/rnatypes/SecStruct.ts:345]: Typo: "opening"

File: src/eterna/rnatypes/SecStruct.ts [hunk 2]
```diff
@@ -333,82 +333,67 @@ export default class SecStruct {
...
+            for (const [bpA, bpB] of this.pairs.entries()) {
+                // Skip unpaired bases (already in our structure) and pairs which map the "closing"
+                // half to the "opening" half (we've already processed this pair from the other direction)
+                if (bpA > bpB || bpB === -1) continue;
+
+                // Find the first pseudoknot degree for which adding this pair would not introduce
+                // a "cross" (ie, 1-10 crosses with 2-11, but not 2-9. We can't use this degree
+                // if there's a cross because then the characters would match with the wrong bases)
+                // We define two pairs (a,b) and (c,d) (with a<b and c<d) as crossed if c>a and
+                // d > b - that is, (a,b) starts, then (c,d) starts, but (a,b) ends before
```
Comment (tkaragianes) [inline src/eterna/rnatypes/SecStruct.ts:354]: Match spacing of earlier phrases?

**Review by tkaragianes — APPROVED** [review]
Updated comments are definitely clearer. I like the addition of example structures to illustrate certain points. I still think the best way to explain this code is walking through an example and showing how the various data structures change as we process the structure, but that's not amenable to code comments. I think this strikes a fine balance of detail and clarity.

**Pull request merged**